sql: fix FK validation join implementation

thoszhang opened the pull request:
This PR updates the SQL query used for VALIDATE CONSTRAINT for foreign keys.
The new implementation is compatible with the recent changes to FK matching
semantics (both MATCH FULL and MATCH SIMPLE). It also uses a merge join, which
will improve performance significantly compared to the old hash join
implementation.

Release note (sql change): VALIDATE CONSTRAINT for foreign keys is now
compatible with the new MATCH FULL and MATCH SIMPLE semantics, and is more
performant.

Fixes #33452

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34365)
<!-- Reviewable:end -->

knz commented:
Thank you for doing this! I will let David and Bram comment on the correctness, but a thought: do you think there's a way to prepare the logical plan for the validation query during the planning phase, so that it would show up in the output of EXPLAIN?

lucy-zhang commented:
@knz probably. But we're planning to move the validation query out of the user transaction and have happen asynchronously (by adding a mutation and moving it through the schema changer), at which point AFAIK it wouldn't be possible. We probably ultimately also want to implement a backfiller instead of using SQL to do this validation as well.

(The future changes I'm describing are implemented for check constraints in #34301)

knz commented:
Thank you. If done asynchronously (in a job I assume) but we're careful to use standard internal executor machinery, the user can still go and inspect the plan using the statement stats page (this shows logical plans now). So that UX interest will be satisifed this way.

vivekmenezes reviewed (changes requested):
Nice work on this. You should probably in a followup PR add this constraint validation to a roachtest that already contains the FK relationship. Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-:-LX_i7SeATJkTdxfQ8_N:b-i40vu1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @dt, and @lucy-zhang)

___
*[pkg/sql/check.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LX_hPAL2nqlmnvE06qK:-LX_hPAL2nqlmnvE06qL:b-ozy99l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd04752ce7e226d91a118cdc12c0cf010111d4d6/pkg/sql/check.go#L77)):*
> ```Go
> }
> 
> func matchFullUnacceptableKeyQuery(
> ```

Can you add a comment about what you're trying to do here?
___
*[pkg/sql/check.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LX_hkn14-ksZnQWIUIq:-LX_hkn14-ksZnQWIUIr:bffbuvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd04752ce7e226d91a118cdc12c0cf010111d4d6/pkg/sql/check.go#L149)):*
> ```Go
> 	}
> 
> 	if srcIdx.ForeignKey.Match == sqlbase.ForeignKeyReference_FULL {
> ```

Can you add a comment here about why you need to run two queries in order to validate the constraint: the one below and the default one. Thanks!
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 1568 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LX_gUdE7qbiQtKILhOg:-LX_gUdE7qbiQtKILhOh:b-8faczs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd04752ce7e226d91a118cdc12c0cf010111d4d6/pkg/sql/logictest/testdata/logic_test/fk#L1568)):*
> ```
> ALTER TABLE b VALIDATE CONSTRAINT fk_ref
> 
> statement ok
> ```

Great tests!


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @dt, @lucy-zhang, and @vivekmenezes)

___
*[pkg/sql/check.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LX_hPAL2nqlmnvE06qK:-LX_ymxQEw9dGLcFG5Hd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd04752ce7e226d91a118cdc12c0cf010111d4d6/pkg/sql/check.go#L77)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can you add a comment about what you're trying to do here?
</blockquote></details>

Done.
___
*[pkg/sql/check.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LX_hkn14-ksZnQWIUIq:-LX_ynbgAxTTL3QYmo6h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd04752ce7e226d91a118cdc12c0cf010111d4d6/pkg/sql/check.go#L149)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can you add a comment here about why you need to run two queries in order to validate the constraint: the one below and the default one. Thanks!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @dt, and @lucy-zhang)



<!-- Sent from Reviewable.io -->

BramGruneir reviewed (changes requested):
This is great!  I'm excited we're finally going to be able to do this.

I've left some comments.

> Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-:-LXtq3zF2bjlK-d18bGt:bnl3nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/check.go, line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXtpPrcEr85Xc95X34d:-LXtpPrcEr85Xc95X34e:b-tal8qw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cc468671745e3e970c182920aa470842a012047/pkg/sql/check.go#L120)):*
> ```Go
> 
> 	return fmt.Sprintf(
> 		`SELECT %s FROM (SELECT * FROM %s@%s WHERE %s) AS s LEFT OUTER JOIN %s@%s AS t ON %s WHERE %s IS NULL LIMIT 1`,
> ```

It would be really great to see a comment with an example or two of how this will look.  With all the placeholders, it's hard to mentally ensure it looks correct.  (and the same for the query from matchFullUnacceptableKeyQuery)

For example, what's the point of the `AS s` and `AS t` if the alias is not used?
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 1382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXtipS8ATpRNaduGPpa:-LXtipS8ATpRNaduGPpb:bjnwz07) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cc468671745e3e970c182920aa470842a012047/pkg/sql/logictest/testdata/logic_test/fk#L1382)):*
> ```
> 
> # Add the constraint separately so that it's unvalidated, so we can test VALIDATE CONSTRAINT.
> statement ok
> ```

Instead of just updating the one test, could you make this two separate tests?

Also, if you could use subtests, it would make this a lot more readable.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 1588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXto74U13GfRvNUK6-k:-LXto74U13GfRvNUK6-l:bcxo4ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cc468671745e3e970c182920aa470842a012047/pkg/sql/logictest/testdata/logic_test/fk#L1588)):*
> ```
> );
> 
> # Add the constraint separately so that it's unvalidated, so we can test VALIDATE CONSTRAINT.
> ```

Again, please just duplicate the test, once with validation and once without.  We need to test both scenarios.  Pretty easy to do with the subtests.


<!-- Sent from Reviewable.io -->

lucy-zhang committed:
sql: fix FK validation join implementation

This PR updates the SQL query used for VALIDATE CONSTRAINT for foreign keys.
The new implementation is compatible with the recent changes to FK matching
semantics (both MATCH FULL and MATCH SIMPLE). It also uses a merge join, which
will improve performance significantly compared to the old hash join
implementation.

Release note (sql change): VALIDATE CONSTRAINT for foreign keys is now
compatible with the new MATCH FULL and MATCH SIMPLE semantics, and is more
performant.

Release note: None

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir and @lucy-zhang)

___
*[pkg/sql/check.go, line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXtpPrcEr85Xc95X34d:-LXuWndv3jN5s_6kP8S_:bcx9pfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cc468671745e3e970c182920aa470842a012047/pkg/sql/check.go#L120)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

It would be really great to see a comment with an example or two of how this will look.  With all the placeholders, it's hard to mentally ensure it looks correct.  (and the same for the query from matchFullUnacceptableKeyQuery)

For example, what's the point of the `AS s` and `AS t` if the alias is not used?
</blockquote></details>

Done. (The aliases get used in the WHERE and ON clauses but it's not the easiest to see... I added a comment)
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 1382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXtipS8ATpRNaduGPpa:-LXuWkKbCafmj8FsAKR7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cc468671745e3e970c182920aa470842a012047/pkg/sql/logictest/testdata/logic_test/fk#L1382)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Instead of just updating the one test, could you make this two separate tests?

Also, if you could use subtests, it would make this a lot more readable.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/fk, line 1588 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXto74U13GfRvNUK6-k:-LXuWlwkBcezgtc1nDwj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cc468671745e3e970c182920aa470842a012047/pkg/sql/logictest/testdata/logic_test/fk#L1588)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Again, please just duplicate the test, once with validation and once without.  We need to test both scenarios.  Pretty easy to do with the subtests.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Just a quick comment on the MATCH FULL check.

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-:-LXzLsY73PNR-U2KWuWc:biu8d7m)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/check.go, line 120 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXtpPrcEr85Xc95X34d:-LXzG5gt90NI76Y-joRG:b-c6lp1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cc468671745e3e970c182920aa470842a012047/pkg/sql/check.go#L120)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Done. (The aliases get used in the WHERE and ON clauses but it's not the easiest to see... I added a comment)
</blockquote></details>

Ah, I can see it now.  Thanks!
___
*[pkg/sql/check.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXzHBlw2jRWq2R0VslP:-LXzHBlw2jRWq2R0VslQ:bqw1c70) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ea4c0adfc6e37e115a5fc4a00a1a69fbcb35fa5/pkg/sql/check.go#L87)):*
> ```Go
> // SELECT * FROM child@c_idx
> // WHERE
> //   NOT (a_id IS NULL AND b_id IS NULL OR a_id IS NOT NULL AND b_id IS NOT NULL)
> ```

This is a little convoluted.

For the null case, you can use COALESCE(col1, col2, col3) IS NULL

And could you add some parentheses to make this a little earlier to read?  Relying on the order of operations is a little tough.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @BramGruneir and @lucy-zhang)

___
*[pkg/sql/check.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-LXzHBlw2jRWq2R0VslP:-LXzQLE94MmhgUdtIUUz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ea4c0adfc6e37e115a5fc4a00a1a69fbcb35fa5/pkg/sql/check.go#L87)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This is a little convoluted.

For the null case, you can use COALESCE(col1, col2, col3) IS NULL

And could you add some parentheses to make this a little earlier to read?  Relying on the order of operations is a little tough.

</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34365#-:-LXzVTZF5-5WIuz8WqlA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang)



<!-- Sent from Reviewable.io -->
